logstreams/src/main/java/io/zeebe/logstreams/impl/log/fs/FsLogStorage.java

```diff
@@ -148,6 +149,64 @@ public long read(
     return opStatus;
   }
 
+  @Override
+  public long readLastBlock(final ByteBuffer readBuffer, final ReadResultProcessor processor) {
+    ensureOpenedStorage();
+
+    final int segmentCount = logSegments.segmentCount;
+    final FsLogSegment lastSegment = logSegments.getSegment(segmentCount - 1);
+
+    if (lastSegment != null && lastSegment.getSizeVolatile() > METADATA_LENGTH) {
+      boolean findLast = true;
+      final int segmentId = lastSegment.getSegmentId();
+      long currentAddress = position(segmentId, METADATA_LENGTH);
+      final int startPosition = readBuffer.position();
+      final int limit = readBuffer.limit();
+      while (findLast) {
+        // re-init
+        readBuffer.position(startPosition);
+        readBuffer.limit(limit);
+
+        final long nextAddress = read(readBuffer, currentAddress, processor);
+
+        if (nextAddress == OP_RESULT_NO_DATA) {
+          findLast = false;
+        } else {
+          if (nextAddress < 0L) {
+            findLast = false;
+          }
+          currentAddress = nextAddress;
+        }
+      }
+
+      return currentAddress;
+    }
+    return OP_RESULT_NO_DATA;
+  }
+
+  @Override
+  public long lookUpApproximateAddress(final long position, LongUnaryOperator positionReader) {
+    final int firstSegmentId = logSegments.getFirst().getSegmentId();
+    final int lastSegmentId = logSegments.getLastSegmentId();
+
+    int segmentFound = firstSegmentId;
+
+    for (int segmentId = firstSegmentId; segmentId <= lastSegmentId; segmentId++) {
+      final long blockAddress = getFirstBlockAddress(segmentId);
+      if (blockAddress >= 0) {
```

Comment:
If we skip a segment, which does not contain the event, it is ok. 
If we skip a segment which contains the event, the next segment has a higher position. Then the `segmentFound` will be the previous segment to the skipped one. So it can still seek to the event by the iterative seek.